Task01/Var1

Dsin72 opened the pull request:
Первое дз
- добавила функцию чтения лабиринта
- добавила функцию парсинга лабиринта
- добавила файл с лабиринтом
- добавила создание матрицы смежности из карты лабиринта

Dsin72 committed:
task01

- добавила функцию чтения лабиринта
- добавила функцию парсинга лабиринта
- добавила файл с лабиринтом

Dsin72 committed:
task01

- Добавила создание матрицы смежности из карты лабиринта

xthebat commented on sinyutina_ds/task01/main.py:
```diff
@@ -0,0 +1,82 @@
+        path = input()
+    elif len(args) > 3:
+        print("слишком много аргументов, введите название файла: ")
+        path = input()
```
не нужно останавливать программу после выдачи ошибки. Вспомните любую линуксовую утилиту, если что-то задали неправильно. Ругается и выходит.
То есть я бы написал так:
```python
if len(args) != 2:
    print("Wrong number of args, just file name required")
    sys.exit(-1)
```

Dsin72 replied:
Исправила в новом коммите

xthebat commented on sinyutina_ds/task01/main.py:
```diff
@@ -0,0 +1,82 @@
+    vertex_id = 1
+
+    adj = dict()
+    vertex_map = [['' for i in range(len(row))] for row in maze_map]
```
Почему id вертексов сделаны строками? Где-то они используются как int, где-то как строка

Dsin72 replied:
Показалось наиболее логичным решением на тот момент. В новой версии переделала.

xthebat commented on sinyutina_ds/task01/main.py:
```diff
@@ -0,0 +1,82 @@
+    adj = dict()
+    vertex_map = [['' for i in range(len(row))] for row in maze_map]
+
+    for i in range(len(maze_map[0])):
```
Если я правильно понял, то здесь выполняется построение матрицы смежности, но вопрос, почему проходится только первый столбец?

Dsin72 replied:
Решила разделить обработку первой строки и всех последующих, тк при обработки первой строки не имеет смысла просмотр соседнего элемента из предыдущей строки.

xthebat commented on sinyutina_ds/task01/main.py:
```diff
@@ -0,0 +1,82 @@
+            vertex_id += 1
+
+        prev = maze_map[0][i]
+
```
Функция явно требует того, чтобы бы разделенной как минимум на две подфункции.

Dsin72 replied:
Исправила в новом коммите

xthebat commented on sinyutina_ds/task01/main.py:
```diff
@@ -0,0 +1,82 @@
+
+        prev = maze_map[0][i]
+
+    for i in range(1, len(maze_map)):
```
Хотя вроде как основная обработка находится в этом месте кода, тогда что делает предыдущий цикл?

Dsin72 replied:
Исправила в новом коммите

xthebat reviewed (changes requested):
Не очень понятно решение с prev как это работает, но вроде потестил - похоже.

Мой вариант:
```python
def rangend(*args: int) -> Iterator:
    return itertools.product(*(range(it) for it in args))


def neighbours(maze: List[List[int]], row: int, col: int):
    rows = len(maze)
    cols = len(maze[0])
    for drow, dcol in ((-1, 0), (1, 0), (0, -1), (0, 1)):
        next_row = row + drow
        next_col = col + dcol
        if 0 <= next_row < rows and 0 <= next_col < cols:
            yield next_row, next_col


def get_or_create_node(nodes, row, col):
    if (row, col) not in nodes:
        nodes[(row, col)] = max(nodes.values()) + 1 if len(nodes) else 0
    return nodes[(row, col)]


def generate_graph(maze: List[List[int]]):
    rows = len(maze)
    cols = len(maze[0])

    nodes = dict()
    adj = dict()

    for row, col in filter(lambda it: maze[it[0]][it[1]], rangend(rows, cols)):
        node_id = get_or_create_node(nodes, row, col)
        links = dict()
        adj[node_id] = links
        for neighbour_row, neighbour_col in filter(lambda it: maze[it[0]][it[1]], neighbours(maze, row, col)):
            neighbour_id = get_or_create_node(nodes, neighbour_row, neighbour_col)
            links[neighbour_id] = 1

    return nodes, adj


def adjacency_matrix(adj):
    size = len(adj)
    matrix = [[0] * size for _ in range(size)]
    for node, links in adj.items():
        row = int(node)
        for other, length in links.items():
            col = int(other)
            matrix[row][col] = length
    return matrix
```

нужны:
- unit-test
- "оптимизация" графа

Dsin72 committed:
task01

- Добавила комментарии к алгоритму построения не оптимизированной матрицы смежности
- Зарефакторила алгоритм построения не оптимизированной матрицы смежности:
-- Убрала цикл предварительной обработки
--  Изменила тип идентификатора вершины на кортеж координат
- Добавила функцию оптимизации графа, построенной на обходе в ширину
- Изменила обработку неверного числа аргументов программы:
-- Теперь программа ругается и завершается со статусом -1 (255) 
- Добавила юнит тесты:
-- Тест на чтение карты лабиринта из файла
-- Тест на создание графа
-- Тест на упрощение ациклического графа
-- Тест на упрощение графа (может содержать циклы)
-- Тест на упрощение графа с начальной вершиной, принадлежащей простому пути
-- Тест на упрощение графа с конечной вершиной, принадлежащей простому пути

Dsin72 committed:
task01

- Добавила поиск пути алгоритмом Дейкстры 
- Добавила типизацию 
- Вынесла код добавления связи матрицы смежности в отдельную функцию
- Заменила словарь visited на множество visited
- Добавила тесты
-- Тест на нахождение кратчайшего пути в графе с одним простым путем
-- Тест на нахождение кратчайшего пути в графе без пути от начала до конца
-- Тест на нахождение кратчайшего пути в обобщенном графе
